File: 3670.maximum-product-of-two-integers-with-no-common-bits.cpp
#
# @lc app=leetcode id=3670 lang=cpp
#
# [3670] Maximum Product of Two Integers With No Common Bits
#
# @lc code=start
class Solution {
public:
    long long maxProduct(vector<int>& nums) {
        long long maxProd = 0;
        int n = nums.size();
        
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                // Check if nums[i] and nums[j] have no common bits
                if ((nums[i] & nums[j]) == 0) {
                    long long prod = (long long)nums[i] * nums[j];
                    maxProd = max(maxProd, prod);
                }
            }
        }
        
        return maxProd;
    }
};
# @lc code=end

File: 3670.maximum-product-of-two-integers-with-no-common-bits.java
#
# @lc app=leetcode id=3670 lang=java
#
# [3670] Maximum Product of Two Integers With No Common Bits
#
# @lc code=start
class Solution {
    public long maxProduct(int[] nums) {
        Map<Integer, Integer> maskToMax = new HashMap<>();
        for (int x : nums) {
            int mask = x;
            maskToMax.put(mask, Math.max(maskToMax.getOrDefault(mask, 0), x));
        }
        long ans = 0;
        Integer[] masks = maskToMax.keySet().toArray(new Integer[0]);
        for (int i = 0; i < masks.length; i++) {
            for (int j = i + 1; j < masks.length; j++) {
                if ((masks[i] & masks[j]) == 0) {
                    long prod = 1L * maskToMax.get(masks[i]) * maskToMax.get(masks[j]);
                    if (prod > ans) ans = prod;
                }
            }
        }
        return ans;
    }
}
# @lc code=end